File: app-android/app/src/main/java/de/tutao/tutanota/push/SseClient.java
package de.tutao.tutanota.push;

import android.util.Log;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import de.tutao.tutanota.Crypto;
import de.tutao.tutanota.NetworkUtils;
import de.tutao.tutanota.Utils;
import de.tutao.tutanota.data.SseInfo;

public class SseClient {
	private static final String TAG = "SSE";
	public static final int RECONNECTION_ATTEMPTS = 3;

	private final Crypto crypto;
	private final NetworkObserver networkObserver;
	private final SseListener sseListener;
	private final SseStorage sseStorage;

	private volatile SseInfo connectedSseInfo;
	private volatile long timeoutInSeconds = 90;
	private int failedConnectionAttempts = 0;

	private final AtomicReference<HttpURLConnection> httpsURLConnectionRef = new AtomicReference<>(null);
	private final LooperThread looperThread = new LooperThread(this::connect);

	SseClient(Crypto crypto, SseStorage sseStorage, NetworkObserver networkObserver, SseListener sseListener) {
		this.crypto = crypto;
		this.networkObserver = networkObserver;
		this.sseListener = sseListener;
		this.sseStorage = sseStorage;

		looperThread.start();

		networkObserver.setNetworkConnectivityListener(connected -> {
			HttpURLConnection connection = httpsURLConnectionRef.get();
			if (connected && connection == null) {
				Log.d(TAG, "ConnectionRef not available, schedule connect because of network state change");
				reschedule(0);
			}
		});
	}

	private void reschedule(int delayInSeconds) {
		if (looperThread.getHandler() != null) {
			looperThread.getHandler().postDelayed(this::connect, TimeUnit.SECONDS.toMillis(delayInSeconds));
		} else {
			Log.d(TAG, "looper thread is starting, skip additional reschedule");
		}
	}

	public void restartConnectionIfNeeded(@NonNull SseInfo sseInfo) {
		SseInfo oldConnectedInfo = this.connectedSseInfo;
		this.connectedSseInfo = sseInfo;

		HttpURLConnection connection = httpsURLConnectionRef.get();
		if (connection == null) {
			Log.d(TAG, "ConnectionRef not available, schedule connect");
			reschedule(0);
		} else if (oldConnectedInfo == null
				|| !oldConnectedInfo.getPushIdentifier().equals(sseInfo.getPushIdentifier())
				|| !oldConnectedInfo.getSseOrigin().equals(sseInfo.getSseOrigin())) {
			// If pushIdentifier or SSE origin have changed for some reason, restart the connect.
			// If user IDs have changed, do not restart, if current user is invalid we have either oldConnectedInfo
			Log.d(TAG, "ConnectionRef available, but SseInfo has changed, call disconnect to reschedule connection");
			connection.disconnect();
		} else {
			Log.d(TAG, "ConnectionRef available, do nothing");
		}
	}

	private void connect() {
		Log.d(TAG, "Starting SSE connection");
		Random random = new Random();
		BufferedReader reader = null;
		if (connectedSseInfo == null) {
			Log.d(TAG, "sse info not available skip reconnect");
			return;
		}

		if (!this.sseListener.onStartingConnection()) {
			return;
		}
		this.timeoutInSeconds = this.sseStorage.getConnectTimeoutInSeconds();
		if (this.timeoutInSeconds == 0) {
			this.timeoutInSeconds = 90;
		}

		ConnectionData connectionData = prepareSSEConnection();
		try {
			HttpURLConnection httpURLConnection = this.openSseConnection(connectionData);
			reader = new BufferedReader(new InputStreamReader(new BufferedInputStream(httpURLConnection.getInputStream())));
			String event;
			Log.d(TAG, "SSE connection established, listening for events");
			boolean notifiedEstablishedConnection = true;
			while ((event = reader.readLine()) != null) {
				handleLine(event);
				if (notifiedEstablishedConnection) {
					this.sseListener.onConnectionEstablished();
					notifiedEstablishedConnection = false;
				}
			}
		} catch (Exception exception) {
			handleException(random, exception, connectionData.userId);
		} finally {
			if (reader != null) {
				try {
					reader.close();
				} catch (IOException ignored) {
				}
			}
			httpsURLConnectionRef.set(null);
		}
	}

	private void handleException(Random random, Exception exception, String userId) {
		HttpURLConnection httpURLConnection = httpsURLConnectionRef.get();
		try {
			// we get not authorized for the stored identifier and user ids, so remove them
			if (httpURLConnection != null && httpURLConnection.getResponseCode() == 403) {
				Log.e(TAG, "not authorized to connect, disable reconnect for " + userId);
				sseListener.onNotAuthorized(userId);
				return;
			}
		} catch (IOException e) {
			// ignore Exception when getting status code.
		}
		int delayBoundary = (int) (timeoutInSeconds * 1.5);
		int delay = (random.nextInt((int) Math.abs(timeoutInSeconds)) + delayBoundary) / 2;

		failedConnectionAttempts++;
		if (failedConnectionAttempts > RECONNECTION_ATTEMPTS) {
			failedConnectionAttempts = 0;
			Log.e(TAG, "Too many failed connection attempts, will try to sync notifications next time system wakes app up");
			sseListener.onStoppingReconnectionAttempts();
		} else if (this.networkObserver.hasNetworkConnection()) {
			Log.e(TAG, "error opening sse, rescheduling after " + delay + ", failedConnectionAttempts: " + failedConnectionAttempts, exception);
			reschedule(delay);
		} else {
			Log.e(TAG, "network is not connected, do not reschedule ", exception);
			sseListener.onStoppingReconnectionAttempts();
		}
	}

	private void handleLine(String line) {
		failedConnectionAttempts = 0;

		if (!line.startsWith("data: ")) {
			Log.d(TAG, "heartbeat");
			return;
		}
		String data = line.substring(6);
		if (data.matches("^[0-9]+$"))
			return;

		if (data.startsWith("heartbeatTimeout:")) {
			timeoutInSeconds = Integer.parseInt(data.split(":")[1]);
			this.sseStorage.setConnectTimeoutInSeconds(timeoutInSeconds);
			sseListener.onConnectionEstablished();
			return;
		}
		this.sseListener.onMessage(data, this.connectedSseInfo);
		Log.d(TAG, "Executing jobFinished after receiving notifications");
	}

	private String requestJson(String pushIdentifier, String userId) {
		JSONObject jsonObject = new JSONObject();
		try {
			jsonObject.put("_format", "0");
			jsonObject.put("identifier", pushIdentifier);
			JSONArray jsonArray = new JSONArray();
			JSONObject userIdObject = new JSONObject();
			userIdObject.put("_id", generateId());
			userIdObject.put("value", userId);
			jsonArray.put(userIdObject);
			jsonObject.put("userIds", jsonArray);
			return URLEncoder.encode(jsonObject.toString(), "UTF-8");
		} catch (JSONException | UnsupportedEncodingException e) {
			throw new RuntimeException(e);
		}
	}

	private String generateId() {
		byte[] bytes = new byte[4];
		crypto.getRandomizer().nextBytes(bytes);
		return Utils.base64ToBase64Url(Utils.bytesToBase64(bytes));
	}

	private ConnectionData prepareSSEConnection() {
		if (connectedSseInfo.getUserIds().isEmpty()) {
			throw new IllegalStateException("Push identifier but no user IDs");
		}
		String userId = connectedSseInfo.getUserIds().iterator().next();

		String json = requestJson(connectedSseInfo.getPushIdentifier(), userId);
		URL url;
		try {
			url = new URL(connectedSseInfo.getSseOrigin() + "/sse?_body=" + json);
		} catch (MalformedURLException e) {
			throw new RuntimeException(e);
		}
		return new ConnectionData(userId, url);
	}

	@NonNull
	private HttpURLConnection openSseConnection(ConnectionData connectionData) throws IOException {
		HttpURLConnection httpsURLConnection = (HttpURLConnection) connectionData.url.openConnection();
		this.httpsURLConnectionRef.set(httpsURLConnection);
		httpsURLConnection.setRequestMethod("GET");
		httpsURLConnection.setRequestProperty("Content-Type", "application/json");
		httpsURLConnection.setRequestProperty("Connection", "Keep-Alive");
		httpsURLConnection.setRequestProperty("Keep-Alive", "header");
		httpsURLConnection.setRequestProperty("Connection", "close");
		httpsURLConnection.setRequestProperty("Accept", "text/event-stream");
		NetworkUtils.addCommonHeaders(httpsURLConnection);

		httpsURLConnection.setConnectTimeout((int) TimeUnit.SECONDS.toMillis(5));
		httpsURLConnection.setReadTimeout((int) (TimeUnit.SECONDS.toMillis(timeoutInSeconds) * 1.2));
		return httpsURLConnection;
	}

	public void stopConnection() {
		HttpURLConnection connection = httpsURLConnectionRef.get();
		Log.d(TAG, "Disconnect sse client");
		if (connection != null) {
			connection.disconnect();
			// check in connect() prevents rescheduling new connection attempts
			this.connectedSseInfo = null;
		}
	}

	public interface SseListener {
		/**
		 * @return {@code true} to continue connecting
		 */
		boolean onStartingConnection();

		/**
		 * Will block reading from SSE until this returns
		 */
		void onMessage(String data, SseInfo sseInfo);

		void onConnectionEstablished();

		void onNotAuthorized(String userId);

		void onStoppingReconnectionAttempts();

	}

	private static final class ConnectionData {
		final String userId;
		final URL url;

		ConnectionData(String userId, URL url) {
			this.userId = userId;
			this.url = url;
		}
	}
}

File: app-android/app/src/main/java/de/tutao/tutanota/push/SseClient.kt
package de.tutao.tutanota.push

import android.util.Log
import de.tutao.tutanota.*
import de.tutao.tutashared.AndroidNativeCryptoFacade
import de.tutao.tutashared.addCommonHeadersWithSysModelVersion
import de.tutao.tutashared.base64ToBase64Url
import de.tutao.tutashared.data.SseInfo
import de.tutao.tutashared.push.SseStorage
import de.tutao.tutashared.toBase64
import okhttp3.*
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.*
import java.net.URL
import java.net.URLEncoder
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference
import kotlin.math.abs

class SseClient internal constructor(
	private val crypto: AndroidNativeCryptoFacade,
	private val sseStorage: SseStorage,
	private val networkObserver: NetworkObserver,
	private val sseListener: SseListener,
	private val defaultClient: OkHttpClient
) {
	@Volatile
	private var connectedSseInfo: SseInfo? = null

	@Volatile
	private var timeoutInSeconds: Long = 90
	private var failedConnectionAttempts = 0
	private val response = AtomicReference<Response?>(null)
	private val looperThread = LooperThread { connect() }
	private fun reschedule(delayInSeconds: Int) {
		if (looperThread._handler != null) {
			looperThread._handler!!.postDelayed({ connect() }, TimeUnit.SECONDS.toMillis(delayInSeconds.toLong()))
		} else {
			Log.d(TAG, "looper thread is starting, skip additional reschedule")
		}
	}

	fun restartConnectionIfNeeded(sseInfo: SseInfo) {
		val oldConnectedInfo = connectedSseInfo
		connectedSseInfo = sseInfo
		val response = response.get()
		if (response == null) {
			Log.d(TAG, "restart requested and connectionRef is not available, schedule connect")
			reschedule(0)
		} else if (
			oldConnectedInfo == null ||
			oldConnectedInfo.pushIdentifier != sseInfo.pushIdentifier ||
			oldConnectedInfo.sseOrigin != sseInfo.sseOrigin
		) {
			// If pushIdentifier or SSE origin have changed for some reason, restart the connect.
			// If user IDs have changed, do not restart, if current user is invalid we have either oldConnectedInfo
			Log.d(
				TAG,
				"restart requested, connectionRef is available, but sseInfo has changed, call disconnect to reschedule connection"
			)
			response.close()
		} else {
			Log.d(TAG, "restart requested but connectionRef available and didn't change, do nothing")
		}
	}

	private fun connect() {
		Log.d(TAG, "Starting SSE connection")
		val random = Random()
		val connectedSseInfo = this.connectedSseInfo
		if (connectedSseInfo == null) {
			Log.d(TAG, "sse info not available skip reconnect")
			return
		}
		if (!sseListener.onStartingConnection()) {
			return
		}
		timeoutInSeconds = sseStorage.getConnectTimeoutInSeconds()
		if (timeoutInSeconds == 0L) {
			timeoutInSeconds = 90
		}
		val connectionData = prepareSSEConnection(connectedSseInfo)
		try {
			var shouldNotifyAboutEstablishedConnection = true
			val response = openSseConnection(connectionData)
			Log.d(TAG, "connected, listening for events, ${response.code} ${response.isSuccessful}")
			if (!response.isSuccessful) {
				handleFailedConnection(random, connectionData.userId, null)
				return
			}

			BufferedReader(InputStreamReader(response.body!!.byteStream())).forEachLine {
				handleLine(it)
				if (shouldNotifyAboutEstablishedConnection) {
					// We expect to get at least one event right away so we don't consider the connection "established"
					// until it happens.
					sseListener.onConnectionEstablished()
					shouldNotifyAboutEstablishedConnection = false
				}
			}
		} catch (exception: Exception) {
			handleFailedConnection(random, connectionData.userId, exception)
		} finally {
			sseListener.onConnectionBroken()
			response.set(null)
		}
	}

	private fun handleFailedConnection(random: Random, userId: String, exception: Exception?) {
		val response = response.get()
		Log.d(TAG, "connection failed, $userId $exception ${response?.code}")
		try {
			// we get not authorized for the stored identifier and user ids, so remove them
			if (response != null && response.code == 401) {
				// unless there is another user id added there is no reason to try to reconnect
				Log.e(TAG, "not authorized to connect, disable reconnect")
				sseListener.onNotAuthorized(userId)
				return
			}
		} catch (e: IOException) {
			// ignore Exception when getting status code.
		}
		val delayBoundary = (timeoutInSeconds * 1.5).toInt()
		val delay = (random.nextInt(abs(timeoutInSeconds).toInt()) + delayBoundary) / 2
		failedConnectionAttempts++
		when {
			failedConnectionAttempts > RECONNECTION_ATTEMPTS -> {
				failedConnectionAttempts = 0
				Log.e(
					TAG,
					"Too many failed connection attempts, will try to sync notifications next time system wakes app up"
				)
				sseListener.onStoppingReconnectionAttempts()
			}
			networkObserver.hasNetworkConnection() -> {
				Log.e(
					TAG,
					"error opening sse, rescheduling after $delay, failedConnectionAttempts: $failedConnectionAttempts",
					exception
				)
				reschedule(delay)
			}
			else -> {
				Log.e(TAG, "network is not connected, do not reschedule ", exception)
				sseListener.onStoppingReconnectionAttempts()
			}
		}
	}

	private fun handleLine(line: String) {
		failedConnectionAttempts = 0
		if (!line.startsWith("data: ")) {
			Log.d(TAG, "heartbeat")
			return
		}
		val data = line.substring(6)
		if (data.matches(Regex("^[0-9]+$"))) return
		if (data.startsWith("heartbeatTimeout:")) {
			timeoutInSeconds = data.split(":".toRegex()).toTypedArray()[1].toInt().toLong()
			sseStorage.setConnectTimeoutInSeconds(timeoutInSeconds)
			return
		}
		sseListener.onMessage(data, connectedSseInfo)
		Log.d(TAG, "onMessage")
	}

	private fun requestJson(pushIdentifier: String, userId: String?): String {
		val jsonObject = JSONObject()
		return try {
			jsonObject.put("_format", "0")
			jsonObject.put("identifier", pushIdentifier)
			val jsonArray = JSONArray()
			val userIdObject = JSONObject()
			userIdObject.put("_id", generateId())
			userIdObject.put("value", userId)
			jsonArray.put(userIdObject)
			jsonObject.put("userIds", jsonArray)
			URLEncoder.encode(jsonObject.toString(), "UTF-8")
		} catch (e: JSONException) {
			throw RuntimeException(e)
		} catch (e: UnsupportedEncodingException) {
			throw RuntimeException(e)
		}
	}

	private fun generateId(): String {
		val bytes = ByteArray(4)
		crypto.randomizer.nextBytes(bytes)
		return bytes.toBase64().base64ToBase64Url()
	}

	private fun prepareSSEConnection(connectedSseInfo: SseInfo): ConnectionData {
		check(!connectedSseInfo.userIds.isEmpty()) { "Push identifier but no user IDs" }
		val userId = connectedSseInfo.userIds.first()
		val json = requestJson(connectedSseInfo.pushIdentifier, userId)
		val url = URL(connectedSseInfo.sseOrigin + "/sse?_body=" + json)
		return ConnectionData(userId, url)
	}

	@Throws(IOException::class)
	private fun openSseConnection(connectionData: ConnectionData): Response {
		val requestBuilder = Request.Builder()
			.url(connectionData.url)
			.method("GET", null)
			.header("Content-Type", "application/json")
			.header("Connection", "Keep-Alive")
			.header("Accept", "text/event-stream")
		addCommonHeadersWithSysModelVersion(requestBuilder)

		val req = requestBuilder.build()

		val response = defaultClient
			.newBuilder()
			.connectTimeout(5, TimeUnit.SECONDS)
			.writeTimeout(5, TimeUnit.SECONDS)
			.readTimeout((timeoutInSeconds * 1.2).toLong(), TimeUnit.SECONDS)
			.build()
			.newCall(req)
			.execute()
		this.response.set(response)
		return response
	}

	fun stopConnection() {
		val response = response.get()
		Log.d(TAG, "Disconnect sse client")
		if (response != null) {
			response.close()
			// check in connect() prevents rescheduling new connection attempts
			connectedSseInfo = null
		}
	}

	interface SseListener {
		/**
		 * @return `true` to continue connecting
		 */
		fun onStartingConnection(): Boolean

		/**
		 * Will block reading from SSE until this returns
		 */
		fun onMessage(data: String, sseInfo: SseInfo?)
		fun onConnectionEstablished()
		fun onConnectionBroken()
		fun onNotAuthorized(userId: String)
		fun onStoppingReconnectionAttempts()
	}

	private class ConnectionData constructor(val userId: String, val url: URL)
	companion object {
		private const val TAG = "SSE"
		const val RECONNECTION_ATTEMPTS = 3
	}

	init {
		looperThread.start()
		networkObserver.setNetworkConnectivityListener { connected ->
			val connection = response.get()
			if (connected && connection == null) {
				Log.d(TAG, "ConnectionRef not available, schedule connect because of network state change")
				reschedule(0)
			}
		}
	}
}